osf/models/mixins.py

```diff
@@ -739,31 +739,67 @@ def subject_hierarchy(self):
             ]
         return []
 
-    def set_subjects(self, new_subjects, auth, add_log=True):
-        """ Helper for setting M2M subjects field from list of hierarchies received from UI.
-        Only authorized admins may set subjects.
+    @property
+    def subjects_relationship_url(self):
+        return self.absolute_api_v2_url + 'relationships/subjects/'
 
-        :param list[list[Subject._id]] new_subjects: List of subject hierarchies to be validated and flattened
-        :param Auth auth: Auth object for requesting user
-        :param bool add_log: Whether or not to add a log (if called on a Loggable object)
+    @property
+    def subjects_url(self):
+        return self.absolute_api_v2_url + 'subjects/'
 
-        :return: None
-        """
+    def check_subject_perms(self, auth):
         AbstractNode = apps.get_model('osf.AbstractNode')
         Preprint = apps.get_model('osf.Preprint')
         CollectionSubmission = apps.get_model('osf.CollectionSubmission')
-        if getattr(self, 'is_registration', False):
-            raise PermissionsError('Registrations may not be modified.')
-        if isinstance(self, (AbstractNode, Preprint)):
+
+        if isinstance(self, AbstractNode):
             if not self.has_permission(auth.user, ADMIN):
                 raise PermissionsError('Only admins can change subjects.')
+        elif isinstance(self, Preprint):
+            if not self.has_permission(auth.user, WRITE):
+                raise PermissionsError('Must have admin or write permissions to change a preprint\'s subjects.')
         elif isinstance(self, CollectionSubmission):
             if not self.guid.referent.has_permission(auth.user, ADMIN) and not auth.user.has_perms(self.collection.groups[ADMIN], self.collection):
                 raise PermissionsError('Only admins can change subjects.')
+        return
+
+    def add_subjects_log(self, old_subjects, auth):
+        self.add_log(
+            action=NodeLog.SUBJECTS_UPDATED,
+            params={
+                'subjects': list(self.subjects.values('_id', 'text')),
+                'old_subjects': list(Subject.objects.filter(id__in=old_subjects).values('_id', 'text'))
+            },
+            auth=auth,
+            save=False,
+        )
+        return
+
+    def assert_subject_format(self, subj_list, expect_list, error_msg):
+        """ Helper for asserting subject request is formatted properly
+        """
+        is_list = type(subj_list) is list
+
+        if (expect_list and not is_list) or (not expect_list and is_list):
+            raise ValidationValueError('Subjects are improperly formatted. {}'.format(error_msg))
+
+    def set_subjects(self, new_subjects, auth, add_log=True):
```

Comment:
Very similar to existing method, which expects subjects to come in as attributes (list of lists). Still maintaining this functionality for backwards compat.